plugins/mongodb/mongodbembed.hpp

```diff
@@ -0,0 +1,696 @@
+                        while (*query && *query != '\n')
+                            query++;
+                    }
+                    else if (*query=='*')
+                    {
+                        query++;
+                        for (;;)
+                        {
+                            if (!*query)
+                                fail("Unterminated comment in query string");
+                            if (*query=='*' && query[1]=='/')
+                            {
+                                query+= 2;
+                                break;
+                            }
+                            query++;
+                        }
+                    }
+                }
+                break;
+            default:
+                if (!inStr && ch==searchFor)
+                    return query;
+                break;
+            }
+        }
+        return NULL;
+    }
+
+    static unsigned countParameterPlaceholders(const char *query)
+    {
+        unsigned queryCount = 0;
+        while ((query = findUnquotedChar(query, '$')) != NULL)
+            queryCount++;
+        return queryCount;
+    }
+
+    enum PathNodeType {CPNTScalar, CPNTDataset, CPNTSet};
+
+    struct PathTracker
+    {
+        StringBuffer nodeName;
+        PathNodeType nodeType;
+        unsigned int currentChildIndex;
+        unsigned int childCount;
+        unsigned int childrenProcessed;
+
+        // Simple constructor
+        PathTracker()
+        {}
+
+        // Constructor given node name and dataset bool
+        PathTracker(const StringBuffer& _nodeName, PathNodeType _nodeType)
+            : nodeName(_nodeName), nodeType(_nodeType), currentChildIndex(0), childCount(0), childrenProcessed(0)
+        {}
+
+        // Copy constructor
+        PathTracker(const PathTracker& other)
+            : nodeName(other.nodeName), nodeType(other.nodeType), currentChildIndex(other.currentChildIndex), childCount(other.childCount), childrenProcessed(other.childrenProcessed)
+        {}
+    };
+
+    /**
+     * @brief Holds information about where to send the query, how to query, and the results of the query.
+     * 
+     */
+    class MongoDBQuery
+    {
+    public:
+        /**
+         * @brief Stores the databaseName and CollectionName for executing the operations.
+         * 
+         * @param database MongoDB database to connect to.
+         * @param collection MongoDB collection to connect to.
+         */
+        MongoDBQuery(const char *database, const char *collection, std::int32_t _batchSize) 
+        {
+            databaseName = database;
+            collectionName = collection;
+            batchSize = _batchSize;
+        }
+
+        /**
+         * @brief Set the Embed object and remove leading characters.
+         * 
+         * @param chars Size of the script.
+         * @param eScript pointer to beginning of the script.
+         */
+        void setEmbed(size32_t &chars, const char *eScript) 
```

Comment:
The chars parameter is confusing.  It isn't really needed since eScript is null terminated in the calling code (since it is assigned to a StringAttr).  If you want to avoid that assignment then you should pass in the size and the original pointer, but the size does not need to be a reference parameter.

Reply:
I just removed the parameter since the script is created using StringAttr making it redundant.